### ISSUE TITLE
Compiler produces an unhelpful error message

### ISSUE THREAD

**Issue by JiaJiaJiang** [ziglang/zig#21808]
### Zig Version

0.14.0-dev.2047+985b13934

### Steps to Reproduce and Observed Output

I'm learning zig and trying to write a test demo, the following test will occur a complier error
```zig
//a simple string struct demo
const std = @import("std");
const testing = std.testing;
const debug = std.debug;
const String = @This();

allocator: std.mem.Allocator,
byteLength: u32 = 0,
slice: ?[*:0]u8 = null,
pub fn from(strs: anytype, allocator: std.mem.Allocator) !*String {
    const place = try allocator.create(String);
    place.* = .{
        .slice = null,
        .byteLength = 0,
        .allocator = allocator,
    };
    try place.set(strs);
    return place;
}
pub fn set(self: *String, strs: anytype) !void {
    if (self.slice) |p| {
        self.allocator.free(p);
        self.slice = null;
    }
    debug.print("set strs:{}\n", .{strs.len});
    var len: u32 = 0;
    inline for (strs) |s| {
        len += s.len;
    }
    debug.print("length:{}\n", .{len});
    var offset: usize = 0;
    const ptr: [*:0]u8 = try self.allocator.allocSentinel(u8, len, 0);
    inline for (0..strs.len) |idx| {
        const s = strs[idx];
        @memcpy(ptr[offset..(offset + s.len)], s[0..s.len]);
        offset += s.len;
    }
    self.slice = ptr;
    self.byteLength = len;
}

pub fn str(self: *String) [:0]u8 {
    if (self.slice) |p| {
        return p[0..self.byteLength :0];
    } else {
        // unreachable;
        var emptyStr = "".*;
        const s: [:0]u8 = &emptyStr;
        return s;
    }
}
pub fn deinit(self: *String) void {
    if (self.slice != null) {
        self.allocator.free(self.slice);
    }
    self.allocator.free(self);
}

test "String_test" {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const allocator = gpa.allocator();
    debug.print("String_test\n", .{});
    const str1 = try String.from(.{ "a", "bcde", "fg" }, allocator);
    debug.print("String: {s}\n", .{str1.slice.?});
    try testing.expectEqualStrings(str1.str(), "abcdefg");
}

```
By running the test command, it just output an error message and exit:
```
C:\Users\xxx\scoop\apps\zig-dev\0.14.0-dev.2047\lib\std\mem.zig:4230:14: error: type '[*:0]u8' does not support field access
    if (slice.len == 0 and std.meta.sentinel(Slice) == null) return &[0]u8{};
        ~~~~~^~~~
```

### Expected Output

It may print the position where I cannot use the related type, or it's a bug, not sure.

**Comment by NicoElbers** [comment]
You are trying to get the length of a many item pointer (with a sentinel). Where as a slice consists of a pointer and a length, a many item pointer consists purely of a raw pointer, there is no length.

The compiler error tries to tell you that a field access (.len) is not possible, because pointers don't have fields. This is completely accurate. 

There is a case to be made that specifically trying to access .len on a pointer gives a special compiler error, but everything is behaving as intended. 


Frankly, I did not read your code, but try to use `[:0]u8`, as this is an actual slice instead of `[*:0]u8`. If this is not possible I think std has a function to get the length of a sentinel terminated array, or simply write your own (by iterating until you find the sentinel)

**Comment by JiaJiaJiang** [comment]
@NicoElbers 
Thank you for your reply. I have tried commenting out that debug line and I get the same error message. This is what confuses me because there is no other place in the code to get the len of [*:0]u8, and I know it cannot be obtained. The error message now makes me unable to determine what is wrong.

**Comment by rohlem** [comment]
As the error message states, the issue is triggered [in `std/mem.zig` line `4230`](https://github.com/ziglang/zig/blob/b1361f237aa6b2584e6dbc43c8ae8dc74394eb64/lib/std/mem.zig#L4230) .
This is called by [`std.mem.Allocator.free`](https://github.com/ziglang/zig/blob/b1361f237aa6b2584e6dbc43c8ae8dc74394eb64/lib/std/mem/Allocator.zig#L308), which you call in `set` as `self.allocator.free(p);`.
Because this is a type error, the Zig compiler currently doesn't provide further call stack information. Because it is caused by an `anytype` argument, this type is actually specified by the caller, so that information would be helpful, however such analysis currently isn't implemented/provided.

As pointed out above, it's generally safer to use slices (which contain length information) over sentinel-terminated pointers-to-many.
The `std.mem.Allocator` interface operates using slices. `free` expects a slice, which is why your call with the unwrapped `self.slice` (which is a pointer-to-many) causes the compile error.

Generally, if you're not sure whether something is a compiler bug or not,
feel free to ask in one of the [community spaces](https://github.com/ziglang/zig/wiki/Community) first, as there are usually more people active and willing to troubleshoot and answer questions about status-quo.
This issue tracker is primarily meant to facilitate development of the Zig language and toolchain itself.

**Comment by NicoElbers** [comment]
As rohelm said, community spaces are better places to ask questions about learning Zig. 

However I do think it is a reasonable proposal to add type checks to the free and destroy functions, so beginners don't struggle with issues like this. The compiler error is quite quite vague if you don't know what to look for.

**Comment by JiaJiaJiang** [comment]
Yes, I thought it was a compiler bug, because the error location given by the compiler was in the standard library, not in my code. There happened to be an "Error message improvement" option in the "new issue" option, so I filed this issue.
